File: src/pipeline/analyze.ts
import type { ProgressEvent, Repo2LearnConfig, RepoContext } from "../types";
import { analyzePrompt } from "../prompts/analyze";
import type { CodexDriver } from "../codex/driver";
import { Cache } from "../util/cache";
import { configFingerprint } from "../config";

export async function runAnalyzeStage(args: {
  ctx: RepoContext; driver: CodexDriver; cfg: Repo2LearnConfig; cache: Cache; onProgress?: (e: ProgressEvent) => void;
}): Promise<string> {
  const { ctx, driver, cfg, cache, onProgress } = args;
  const key = cache.key({ stage: "analyze", sha: ctx.sha, cfg: configFingerprint(cfg), v: 4 });
  const cached = await cache.get<string>(key);
  if (cached) { onProgress?.({ type: "log", level: "info", message: "analyze cache hit" }); return cached; }
  const res = await driver.run({ label: "analyze", prompt: analyzePrompt(ctx), cwd: ctx.localPath });
  await cache.set(key, res.text);
  return res.text;
}

File: backend/app/services/pipeline/analyze.py
from __future__ import annotations

from pathlib import Path
from typing import Any

from app.core.config import Settings
from app.services.cache import Cache
from app.services.pipeline.call import CodexDriverLike, codex_json
from app.services.pipeline.prompts import analyze_prompt
from app.services.repo import RepoContext


async def run_analyze_stage(
    *,
    ctx: RepoContext,
    driver: CodexDriverLike,
    cache: Cache,
    settings: Settings,
) -> dict[str, Any]:
    key = cache.key(
        {
            "stage": "analyze-v1",
            "repo": ctx.url,
            "sha": ctx.sha,
            "model": settings.r2l_codex_model,
            "effort": settings.r2l_codex_reasoning_effort,
        }
    )
    cached = cache.get(key)
    if cached is not None:
        return dict(cached)

    analysis = await codex_json(
        driver=driver,
        label="analyze",
        prompt=analyze_prompt(ctx),
        cwd=Path(ctx.localPath),
        model=dict[str, Any],
    )
    cache.set(key, analysis)
    return analysis

